developer-hub/src/reference_docs/trait_based_programming.rs

```diff
@@ -0,0 +1,226 @@
+//! that can be transferred.
+//!
+//! We can foremost write this as simple as the following snippet:
+#![doc = docify::embed!("./src/reference_docs/trait_based_programming.rs", basic)]
+//!
+//!
+//! In this example, we use arbitrary choices for `AccountId`, `Balance` and the `MinTransfer` type.
+//! This works great for **one team's purposes** but we have to remember that Substrate and FRAME
+//! are written as generic frameworks, intended to be highly configurable.
+//!
+//! In a broad sense, there are two avenues in exposing configurability:
+//!
+//! 1. For *values* that need to be generic, for example `MinTransfer`, we attach them to the
+//!    `Pallet` struct as fields:
+//!
+//! ```
+//! struct Pallet {
+//! 	min_transfer: u128,
+//! }
+//! ```
+//!
+//! 2. For *types* that need to be generic, we would have to use generic or associated types, such
+//!    as:
+//!
+//! ```
+//! struct Pallet<AccountId> {
+//! 	min_transfer: u128,
+//!     _marker: std::marker::PhantomData<AccountId>,
+//! }
+//! ```
+//!
+//! Substrate and FRAME, for various reasons (performance, correctness, type safety) has opted to
+//! use *types* to declare both *values* and *types* as generic. This is the essence of why the
+//! `Get` trait exists.
+//!
+//! This would bring us to the second iteration of the pallet, which would look like:
+#![doc = docify::embed!("./src/reference_docs/trait_based_programming.rs", generic)]
+//!
+//! In this example, we managed to make all 3 of our types generic. Taking the example of the
+//! `AccountId`, one should read the above as following:
+//!
+//! > The `Pallet` does not know what type `AccountId` concretely is, but it knows that it is
+//! > something that adheres to being `From<[u8; 32]>`.
+//!
+//! This method would work, but it suffers from two downsides:
+//!
+//! 1. It is verbose, each `impl` block would have to reiterate all of the trait bounds.
+//! 2. It cannot easily share/inherit generic types. Imagine multiple pallets wanting to be generic
+//!    over a single `AccountId`. There is no easy way to express that in this model.
+//!
+//! Finally, this brings us to using traits and associated types on traits to express the above.
+//! trait associated types have the benefit of:
+//!
+//! 1. Being less verbose, as in effect they can *group multiple `type`s together*.
+//! 2. Can inherit from one another by declaring super-trait
+//!
+//! > Interestingly, one of the only downside of associated types is that declaring defaults on them
+//! > is not stable yet. In the meantime, we have built our own custom mechanics around declaring
+//! > defaults for associated types, see [`pallet_default_config_example`].
+//!
+//! The last iteration of our code would look like this:
+#![doc = docify::embed!("./src/reference_docs/trait_based_programming.rs", trait_based)]
+//!
+//! Notice how instead of having multiple generics, everything is generic over a single `<T:
+//! Config>`, and all types are fetched through `T::Foo`.
```

Comment:
I've hesitated here, what is `T::Foo`?
```suggestion
//! Config>`, and all types are fetched through `T`, for example `T::AccountId`, `T::MinTransfer`.
```